archive/imo/imo1962_q4.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2020 Kevin Lacker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Lacker
+-/
+import data.real.pi
+
+open real
+noncomputable theory
+
+notation `π` := real.pi
+
+/-!
+Solve the equation `cos x ^ 2 + cos (2 * x) ^ 2 + cos (3 * x) ^ 2 = 1`.
+
+Since Lean does not have a concept of "simplest form", we just express what is
+in fact the simplest form of the set of solutions, and then prove it equals the set of solutions.
+-/
+
+def problem_equation (x : ℝ) : Prop := cos x ^ 2 + cos (2 * x) ^ 2 + cos (3 * x) ^ 2 = 1
+
+def solution_set : set ℝ :=
+{ x : ℝ | ∃ k : ℤ, x = (2 * ↑k + 1) * π / 4 ∨ x = (2 * ↑k + 1) * π / 6 }
+
+/-
+The key to solving this problem simply is that we can rewrite the equation as
+a product of terms, shown in `alt_formula`, being equal to zero.
+-/
+
+def alt_formula (x : ℝ) : ℝ := cos x * (cos x ^ 2 - 1/2) * cos (3 * x)
+
+lemma cosx2 {x : ℝ} : cos x ^ 2 = 1 - sin x ^ 2 :=
+begin
+  have h1 : sin x ^ 2 = 1 - cos x ^ 2, from real.sin_square x,
+  rw h1,
+  ring
+end
+
+lemma cos3x {x : ℝ} : cos (3 * x) = 4 * cos x ^ 3 - 3 * cos x :=
```

Comment:
`cos (3 * x)`, `sin (3 * x)`, `cosh (3 * x)` and `sinh (3 * x)`, all of them for both real and complex arguments, definitely belong in mathlib proper, as standard results that can be found in any number of sources.

Reply:
(Though I see mathlib is currently missing `cosh_two_mul` and `sinh_two_mul`, so right now the API for hyperbolic functions is less-developed than that for trigonometric functions even when the results correspond more or less exactly.)

Reply:
All right, I added all of these to mathlib, as well as another one or two trig identities that are handy to prove these.